Fix sandboxed iframes to load correct content

TimvdLippe opened the pull request:
There was a discrepancy that for iframes with a sandbox, the script_thread wouldn't handle the `about:` URLs correctly. Now, we reuse the same logic as used when navigating in an already present child frame.

Fixes #36529
Fixes #37499

TimvdLippe committed:
Fix sandboxed iframes to load correct content

There was a discrepancy that for iframes with a sandbox,
the script_thread wouldn't handle the `about:` URLs correctly.
Now, we reuse the same logic as used when navigating in
an already present child frame.

Fixes #36529
Fixes #37499

Signed-off-by: Tim van der Lippe <tvanderlippe@gmail.com>

TimvdLippe reviewed (commented):
See the comments in order to read through the code. While it avoids the CSP error, it doesn't actually load the iframe and I don't understand why. @jdm any thoughts on where the sandbox could have an effect? I feel like we are missing some sort of "hey, go start doing things now" in the iframe. Since it seems like it correctly navigates to the URL and then simply shows a blank screen and does nothing.

jdm commented:
Weirdly, when https://github.com/servo/servo/blob/e1c037815c418dad1a95517c32c09c993ac4b572/components/script/dom/htmliframeelement.rs#L272 from the iframe's post-connection steps, the URL we try to load is about:blank. This suggests to me that the iframe's URL is changed later, but maybe this JS exception prevents the code from running that changes the URL:
```
(new Error("Minified React error #421; visit https://reactjs.org/docs/error-decoder.html?invariant=421 for the full message or use the non-minified dev environment for full errors and additional helpful warnings.", "https://developer.mozilla.org/static/js/main.a918a4e7.js", 2))
```

TimvdLippe commented:
I created a dummy HTML page to test this with a static `src`. However, I don't think it worked there either. Maybe you can use that to test it out? See the second details contents in https://github.com/servo/servo/issues/37499#issuecomment-2985375658

TimvdLippe commented:
The results of `/html/infrastructure/urls/terminology-0/document-base-url-about-srcdoc.https.window.html` look similar to the issue with MDN. It incorrectly leads to loading `about:srcdoc` instead of the WPT url. Need to figure out why that's the case.

jdm commented:
> The results of `/html/infrastructure/urls/terminology-0/document-base-url-about-srcdoc.https.window.html` look similar to the issue with MDN. It incorrectly leads to loading `about:srcdoc` instead of the WPT url. Need to figure out why that's the case.

If you're referring to these failures:
```
  FAIL allow-same-origin - assert_equals: expected "https://web-platform.test:8443/html/infrastructure/urls/terminology-0/document-base-url-about-srcdoc.https.window.html" but got "about:srcdoc"
runTest/<@https://web-platform.test:8443/html/infrastructure/urls/terminology-0/document-base-url-about-srcdoc.https.window.js:20:18
async*Test.prototype.step@https://web-platform.test:8443/resources/testharness.js:2684:25
promise_test/tests.promise_tests</<@https://web-platform.test:8443/resources/testharness.js:738:36
promise_test/tests.promise_tests<@https://web-platform.test:8443/resources/testharness.js:737:20
  FAIL disallow-same-origin - assert_equals: expected "https://web-platform.test:8443/html/infrastructure/urls/terminology-0/document-base-url-about-srcdoc.https.window.html" but got "about:srcdoc"
runTest/<@https://web-platform.test:8443/html/infrastructure/urls/terminology-0/document-base-url-about-srcdoc.https.window.js:20:18
async*Test.prototype.step@https://web-platform.test:8443/resources/testharness.js:2684:25
promise_test/tests.promise_tests</<@https://web-platform.test:8443/resources/testharness.js:738:36
promise_test/tests.promise_tests<@https://web-platform.test:8443/resources/testharness.js:737:20
```
That comes from the document being initialized with https://github.com/servo/servo/blob/9a0f2be162f8f8b3d9fe77cf902b0a6f9dd9729f/components/script/script_thread.rs#L3853-L3854 via https://github.com/servo/servo/blob/9a0f2be162f8f8b3d9fe77cf902b0a6f9dd9729f/components/script/script_thread.rs#L3256 .

jdm commented:
Ooh, and https://github.com/servo/servo/blob/9a0f2be162f8f8b3d9fe77cf902b0a6f9dd9729f/components/script/dom/document.rs#L814-L829 is broken when we load a sandboxed about:blank/about:srcdoc in another script thread, because we can't access the parent browsing context.

TimvdLippe committed:
Add document about base URL

This is used in the fallback base URL for about:blank
and about:srcdoc documents.

Signed-off-by: Tim van der Lippe <tvanderlippe@gmail.com>

TimvdLippe commented:
Pushed my attempt on realigning the implementation to the spec. Unfortunately it doesn't work yet, as the URL is `None` in the case of `about:srcdoc` in a sandbox. In https://github.com/servo/servo/issues/37775#issuecomment-3016464343 I wrote my attempts on figuring out why that's the case, but unfortunately have to give up as I can't find any references to the spec about this.

jdm commented:
>unfortunately have to give up as I can't find any references to the spec about this.

This isn't very surprising to me—this part of the specification (including the lines you quoted in the issue) is much newer than most of the navigation-related code in Servo.

jdm commented:
A lot of the network code was written alongside the original Fetch specification (and evolved from a non-Fetch implementation), and there have been a lot of changes to the specification since that implementation. Likewise, the constellation is one of the earliest pieces of Servo (I remember describing it in technical talks in 2013), and encompasses some complex parts of the specification like session history that have also undergone extensive revisions in the past five years, while the constellation implementation has largely remained unchanged. 

My experience is that navigation, session history, and async fetching are already very complex parts of the specification to reason about, and what exists in the specification has long evolved in a different direction than Servo's implementation (despite our influence like https://github.com/whatwg/html/issues/1454#issuecomment-241609398). This makes it particularly unapproachable if you're used to using the specification text to help guide your understanding of Servo's code, which is a much easier experience in must the script crate (as you have found!).

https://github.com/servo/servo/pull/33087 is an example of my experiments to bring more of our iframe loading implementation in line with the specification. It's a lot of trial and error to figure out where are the right places to make changes, unfortunately. I don't have good suggestions right now, but I may try writing down some high level documentation about what Servo does today so it's easier to see the big picture.

jdm commented:
In particular, https://github.com/whatwg/html/pull/6315 was a _huge_ change to the specification around navigation and session history and occurred while Servo was mostly dormant. As a result we haven't dedicated much effort to adjusting our implementation to catch up with it, leading to the current difficulties of determining how to make correctness changes to Servo's implementation.

TimvdLippe commented:
Closing this in favor of #39610 CC @mrobinson

mrobinson commented:
@TimvdLippe Oh, I didn't realize these two changes were conflicting. Apologies for stepping on your toes here.
